### ISSUE TITLE
Questionable benchmark client

### ISSUE THREAD

**Issue by alexhultman** [now-examples/ws-servers#2]
Hi,

I see that you have benchmarked uws. That's nice and all but there are some things that bother me.

The client is `ws`. To benchmark a server you need to benchmark *the server*, not the client. You cannot show the performance of a high performance C++ server using a low performance scripted server. Consider this flow:

Endpoint (ws client) <------------ network ------------> Endpoint (uws server)

Timer starts client side, client formats a message and sends it (ws -> Node.js -> libuv -> Linux kernel), server gets it (Linux kernel -> uws), parses it, formats it, sends it back (uws -> Linux kernel) and client receives it (Linux kernel -> libuv -> Node.js -> ws) parses it, formats it and sends it back or stops the timer.

If this flow is supposed to benchmark *the server*, what happens client side has to be so optimized and fast that it doesn't taint the result. That means the client *has to* perform its part of the flow *faster* than the server, so that the server is properly stressed. If the client does not have the performance to properly stress the server to 100% CPU time throughout the test then you are benchmarking something other than the server. It can be the network that is the bottleneck, it can be the client, or it can be both.

As you can see from my very simple flow explanation, you see that uws has a far slimmer TCP stack. In that described flow of data, obviously the client will be the bottleneck (we already know ws is majorly underperforming compared to uws), or at least be a significant tainting factor. As a reference, when I benchmark uws using a similar test but with an optimized C++ TCP client directly implemented ontop of libuv, I get 1100 echoes/ms in uws and 30 echoes/ms in ws. As you can see, ws is majorly underperforming compared to uws. This very small difference between ws and uws in your graph is simply because when using uws, you only get *one* endpoint slow. But that one remaining endpoint that is still bottlenecked by ws is going to taint the result and skew the results majorly. You are not seeing the performance of uws, you are only seeing the partial removal of ws's underperformance.

If you want to benchmark the servers in a proper way where they all get to shine with their own true colors then you need to reimplement that benchmarking client in a much more efficient way. And if the networking itself is a bottleneck (if the server is not pinned at 100% CPU time) then you probably need to run this on localhost to remove any networking bottlenecks. The key idea is that you cannot benchmark a sports car by only pushing down the gas pedal 10% of what can be done. uws is not just marginally faster than ws, it is significantly faster (easily up to 20x if you properly stress it).

Hope this criticism gets accepted and acknowledged.

**Comment by alexhultman** [comment]
TLDR; in wsbench.js you require ws as the client. This makes all tests invalid since ws cannot stress uws. By swapping require('ws') in wsbench.js with require('uws') I get some 400% better perf reported for *the server*. In other words, your current results are all majorly invalid since they are capped by ws. This is pretty sad to see here. Please fix. @TooTallNate @rauchg

**Comment by alexhultman** [comment]
Benchmarking **ws** using **ws** as client:
```
node wsbench.js ws://localhost:3000 1000 100
connecting to "ws://localhost:3000"
1000 round trips per client
100 client connections
done!
mean:   2303.63
median: 2303
mode:   2305
range:  48
```

Benchmarking **uws** using **ws** as client:
```
node wsbench.js ws://localhost:3000 1000 100
connecting to "ws://localhost:3000"
1000 round trips per client
100 client connections
done!
mean:   2322.02
median: 2323
mode:   2314
range:  18
```

Benchmarking **uws** using **uws** as the client:
```
node wsbench.js ws://localhost:3000 1000 100
connecting to "ws://localhost:3000"
1000 round trips per client
100 client connections
done!
mean:   484.4
median: 489
mode:   445
range:  82
```

Benchmarking **ws** using **uws** as the client:
```
node wsbench.js ws://localhost:3000 1000 100
connecting to "ws://localhost:3000"
1000 round trips per client
100 client connections
done!
mean:   1920.55
median: 1920
mode:   1920
range:  24
```

As you can **obviously see** the results are **seriously capped by ws** client performance. uws gets a 4-5x boost by being benchmarked by **something that can actually stress it**. *facepalm*
